### PR TITLE
feat(transport): add OpenTelemetry context propagation

### PR DESCRIPTION
**Pull request by quartzmo**
## OpenCensus sunsetting and OpenTelemetry support

The OpenCensus project is obsolete and was archived on July 31st, 2023. This means that any security vulnerabilities that are found will not be patched. We recommend that you begin migrating to OpenCensus tracing to OpenTelemetry, the successor project.

Using the [OpenTelemetry-Go - OpenCensus Bridge](https://github.com/open-telemetry/opentelemetry-go/tree/main/bridge/opencensus), you can immediately begin exporting your traces with OpenTelemetry, even while dependencies such as the Google Cloud client libraries for Go remain instrumented with OpenCensus. If you do not use the bridge, you will need to migrate your entire application and all of its instrumented dependencies at once.  For simple applications, this may be possible, but we expect the bridge to be helpful if multiple libraries with instrumentation are used.

One year after the Google Cloud Go client libraries release the experimental, opt-in OpenTelemetry tracing support in this PR (mid-Q4 2023), the prior experimental OpenCensus tracing will be removed.

Please refer to the following resources:

* [Sunsetting OpenCensus](https://opentelemetry.io/blog/2023/sunsetting-opencensus/)
* [OpenTelemetry-Go - OpenCensus Bridge](https://github.com/open-telemetry/opentelemetry-go/tree/main/bridge/opencensus)
* [What is OpenTelemetry?](https://opentelemetry.io/docs/what-is-opentelemetry/)
* [Cloud Trace - Go and OpenTelemetry](https://cloud.google.com/trace/docs/setup/go-ot)
* On GCE, use [Ops Agent and OpenTelemetry](https://cloud.google.com/trace/docs/otlp)

## OpenTelemetry trace context propagation

The context propagation support for OpenCensus tracing (based on the Cloud Trace header) must be supplemented with new and separate context propagation support for OpenTelemetry tracing (based on the W3C Trace Context header). The `google.golang.org/api/transport` package currently configures context propagation for the Cloud Trace header (X-Cloud-Trace-Context). This is used in all Go client libraries, both HTTP and gRPC, Discovery (Apiary) and GAPIC.

This PR adds support for  by wrapping the underlying transport as follows:

```go
func addOpenTelemetryTransport(trans http.RoundTripper, settings *internal.DialSettings) http.RoundTripper {
	if settings.TelemetryDisabled {
		return trans
	}
	return otelhttp.NewTransport(trans)
}
```

### Passing options to otelhttp.NewTransport in HTTP-based clients

Users who want to pass options to `otelhttp.NewTransport` can disable the library support in the transport package and configure the transport and client themselves.

```go
ctx := context.Background()
trans, err := htransport.NewTransport(ctx,
        http.DefaultTransport,
        option.WithScopes(storage.ScopeFullControl),
)
if err != nil {
        log.Fatal(err)
}
// An example of passing options to the otelhttp.Transport.
otelOpts := otelhttp.WithFilter(func(r *http.Request) bool {
	return r.URL.Path != "/ping"
})
hc := &http.Client{
	Transport: otelhttp.NewTransport(trans, otelOpts),
}
client, err := storage.NewClient(ctx, option.WithHTTPClient(hc))
```

Note that scopes must be set manually in the user-configured solution.

###  Passing options to otelgrpc.NewClientHandler in gRPC-based clients

The setup is pretty similar in gRPC using the [otelgrpc.NewClientHandler](https://pkg.go.dev/go.opentelemetry.io/contrib/instrumentation/google.golang.org/grpc/otelgrpc#NewClientHandler).

```go
projectID := "..."
ctx := context.Background()

// An example of passing options to grpc.WithStatsHandler.
otelOpts := otelgrpc.WithMessageEvents(otelgrpc.ReceivedEvents)
dialOpts := grpc.WithStatsHandler(otelgrpc.NewClientHandler(otelOpts))

ctx := context.Background()
c, err := datastore.NewClient(ctx, projectID, option.WithGRPCDialOption(dialOpts))
if err != nil {
	log.Fatal(err)
}
defer c.Close()
```

### CODE REVIEW

**Comment by tmc** [comment]
As most of the ecosystem has moved on to opentelemetry, it'd be interesting to hear how otel support here is going to evolve (and what the deprecation path of the opencensus instrumentation looks like). Any input there?

**Review by codyoss — APPROVED** [review]

**Review by shollyman — APPROVED** [review]
LGTM, modulo a bit of feedback on the dialcontext testing.

File: transport/grpc/dial_test.go [hunk 1]
```diff
@@ -14,8 +14,21 @@ import (
 	"cloud.google.com/go/compute/metadata"
 	"golang.org/x/oauth2/google"
 	"google.golang.org/api/internal"
+	"google.golang.org/grpc"
 )
 
+func TestDial(t *testing.T) {
+	dialContext = func(ctxGot context.Context, target string, opts ...grpc.DialOption) (conn *grpc.ClientConn, err error) {
```
[the remark this answers is not in the document]
Comment (quartzmo) [reply transport/grpc/dial_test.go:21]: Good call, will do.

[the remark this answers is not in the document]
Comment (quartzmo) [reply transport/grpc/dial_test.go:23]: Fixed, thanks!

**Pull request merged**